core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -222,4 +230,89 @@ public ByteSizeValue paramAsSize(String key, ByteSizeValue defaultValue) {
         return params;
     }
 
+    /**
+     * Does this request have content or a {@code source} parameter? Use this instead of {@link #hasContent()} if this
+     * {@linkplain RestHandler} treats the {@code source} parameter like the body content.
+     */
+    public final boolean hasContentOrSourceParam() {
```

Comment:
I'd kind of prefer to drop this in favor of `contentOrSourceParamParserOrNull` because of the nice dichotomy between it and `contentOrSourceParamParser` but that'd be a larger.